File: tools/run_local.py
# tools/run_local.py
#
# Kernel-enabled harness for JQS + SnapSpace Lite (Replit).
#
# Runs:
# - Evidence registry
# - Manual juror paste
# - JQS decision + receipt
# - OPTIONAL SnapSpace kernel commit (exactly-once, monotonic, replay-proof)
#
# Output shows:
# - JQS verdict
# - Kernel outcome: COMMIT / DENY / FREEZE
# - Signed receipt
# - Saved artifacts + audit log

from __future__ import annotations

import os
import sys
import json
from typing import List

sys.path.append(os.path.dirname(os.path.dirname(__file__)))

from core.evidence import EvidenceRegistry, EvidenceType
from core.contract import build_juror_prompt
from core.orchestrator import run_session, JurorInput
from core.signer import verify_receipt
from core.kernel import SnapSpaceLiteKernel, Attempt


def _read_multiline(prompt: str) -> str:
    print(prompt)
    print("(Paste JSON. Type END on its own line to finish.)")
    lines: List[str] = []
    while True:
        try:
            line = input()
        except EOFError:
            break
        if line.strip() == "END":
            break
        lines.append(line)
    return "\n".join(lines).strip()


def main() -> None:
    case_id = "demo_case"
    required_predicates = ["REPLAY_SAFE", "ROLLBACK_SAFE", "CITATION_GATING_WORKS"]

    signer_kid = "jqs-k1"
    signer_key = b"dev-secret-change-me"  # demo only

    audit_path = os.path.join("audit", "jqs_audit.jsonl")

    # --- Kernel on ---
    kernel = SnapSpaceLiteKernel(state_dir="snapspace_state", spent_window=2048)

    # You control the timeline (monotonic). Start at 0/0 and increment each run.
    print("\n=== Kernel Attempt ===")
    epoch = int(input("epoch (int, e.g. 0): ").strip() or "0")
    turn = int(input("turn  (int, e.g. 0 then 1 then 2...): ").strip() or "0")

    # Define the irreversible action we want to gate.
    # Keep it boring, deterministic, and minimal.
    action = {
        "action_type": "DEMO_COMMIT",
        "case_id": case_id,
        "epoch": epoch,
        "turn": turn,
    }

    # --- Evidence ---
    reg = EvidenceRegistry(allow_payload_share=False)
    ev1 = reg.register(
        evidence_type=EvidenceType.TEXT,
        payload=b"Replay protection check (demo evidence).",
        meta={"name": "replay_log", "source": "demo"},
    )
    ev2 = reg.register(
        evidence_type=EvidenceType.TEXT,
        payload=b"Rollback protection check (demo evidence).",
        meta={"name": "rollback_log", "source": "demo"},
    )
    evidence_index = reg.export_index()

    print("\n=== Evidence Registered ===")
    print("Evidence root hash:", evidence_index["root_hash"])
    print("Evidence IDs:")
    print(" -", ev1.evidence_id, ev1.meta.get("name"))
    print(" -", ev2.evidence_id, ev2.meta.get("name"))

    # --- Prompt ---
    from core.predicates import default_predicates
    prompt = build_juror_prompt(
        role="GENERAL_JUROR",
        case_id=case_id,
        predicates=default_predicates(),
        evidence_index=evidence_index,
    )
    print("\n=== Juror Prompt (paste into LLM juror) ===\n")
    print(prompt)

    # --- Collect juror outputs ---
    jurors: List[JurorInput] = []
    print("\n=== Paste Juror Outputs ===")
    print("Provide one or more jurors. Leave juror ID empty to finish.\n")

    while True:
        juror_id = input("Juror ID (e.g. gpt, gemini, human1) [enter to finish]: ").strip()
        if not juror_id:
            break
        raw = _read_multiline(f"\nPaste output for juror '{juror_id}':")
        jurors.append(JurorInput(juror_id=juror_id, raw_text=raw))
        print()

    if not jurors:
        print("\nNo juror outputs provided. Exiting.")
        return

    # --- Run session with kernel commit ---
    from core.predicates import required_predicate_ids, default_predicates
    preds = default_predicates()
    res = run_session(
        case_id=case_id,
        required_predicates=required_predicate_ids(preds),
        juror_inputs=jurors,
        evidence_registry=reg,
        signer_kid=signer_kid,
        signer_key=signer_key,
        audit_log_path=audit_path,
        force_escalate=False,
        kernel=kernel,
        attempt=Attempt(epoch=epoch, turn=turn),
        action=action,
    )

    print("\n=== JQS Result ===")
    print("Verdict:", res.verdict.value)
    print("Blocking predicates:", res.blocking_predicates)
    print("Predicate results:")
    for k, v in res.predicate_results.items():
        print(f" - {k}: {v.value}")

    print("\n=== Kernel Result ===")
    if res.kernel:
        print(f"Outcome: {res.kernel.outcome}")
        print(f"Reason:  {res.kernel.reason_code}")
        print(f"Epoch:   {res.kernel.last_epoch}")
        print(f"Turn:    {res.kernel.last_turn}")
        print(f"Head:    {res.kernel.head_hash[:16]}...")
    else:
        print("(Kernel not enabled)")

    # Reconstruct unsigned payload the same way as orchestrator did
    unsigned_payload = res.unsigned_output

    print("\n=== Signed Receipt ===")
    print("Unsigned payload:", json.dumps(unsigned_payload, indent=2, sort_keys=True))
    print("receipt_sig:", res.receipt)

    ok = verify_receipt(
        unsigned_output=unsigned_payload,
        receipt_sig=res.receipt,
        kid_to_key={signer_kid: signer_key},
    )
    print("\nReceipt verifies:", ok)

    # Save artifacts
    os.makedirs("out", exist_ok=True)
    unsigned_path = os.path.join("out", "unsigned.json")
    sig_path = os.path.join("out", "receipt_sig.json")
    with open(unsigned_path, "w", encoding="utf-8") as f:
        json.dump(unsigned_payload, f, sort_keys=True, indent=2)
    with open(sig_path, "w", encoding="utf-8") as f:
        json.dump(res.receipt, f, sort_keys=True, indent=2)

    print("\nSaved artifacts:")
    print(" -", unsigned_path)
    print(" -", sig_path)
    print("Audit log:", audit_path)
    print("Kernel state dir: snapspace_state/")
    print("Done.")


if __name__ == "__main__":
    main()

File: tools/run-local.ts
/**
 * tools/run-local.ts
 *
 * Runnable harness for JQS v0.
 *
 * What it does:
 * - Creates an EvidenceRegistry and registers a couple evidence blobs
 * - Defines required predicates
 * - Prints a juror prompt you can paste into an LLM
 * - Lets you paste juror JSON responses back in (manual workflow)
 * - Runs the session orchestrator
 * - Prints verdict + signed receipt
 * - Writes audit log to audit/jqs_audit.jsonl
 *
 * How to run:
 *   npx tsx tools/run-local.ts
 *
 * Note:
 * - No APIs required
 * - This is a manual "human-in-the-loop" juror workflow
 */

import * as readline from "readline";
import * as fs from "fs";
import * as path from "path";
import { EvidenceRegistry, EvidenceType } from "../server/evidence";
import { buildJurorPrompt } from "../server/contract";
import { runSession, JurorInput } from "../server/orchestrator";
import { verifyReceipt } from "../server/signer";

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

function ask(question: string): Promise<string> {
  return new Promise((resolve) => {
    rl.question(question, (answer) => {
      resolve(answer);
    });
  });
}

async function readMultiline(prompt: string): Promise<string> {
  console.log(prompt);
  console.log("(Paste JSON. Type END on its own line to finish.)");

  const lines: string[] = [];

  return new Promise((resolve) => {
    const listener = (line: string) => {
      if (line.trim() === "END") {
        rl.removeListener("line", listener);
        resolve(lines.join("\n").trim());
      } else {
        lines.push(line);
      }
    };
    rl.on("line", listener);
  });
}

async function main(): Promise<void> {
  // --- Config ---
  const caseId = "demo_case";
  const requiredPredicates = [
    "REPLAY_SAFE",
    "ROLLBACK_SAFE",
    "CITATION_GATING_WORKS",
  ];

  // Demo signer (DO NOT use this key in real life)
  const signerKid = "jqs-k1";
  const signerKey = Buffer.from("dev-secret-change-me", "utf-8");

  const auditPath = "audit/jqs_audit.jsonl";

  // --- Evidence Registry ---
  const reg = new EvidenceRegistry({ allowPayloadShare: false });

  // Add some demo evidence items (the content doesn't matter; IDs do)
  const ev1 = reg.register({
    evidenceType: EvidenceType.TEXT,
    payload: Buffer.from("Test log: replay prevention check passed in prior run."),
    meta: { name: "replay_test_log", source: "demo" },
  });
  const ev2 = reg.register({
    evidenceType: EvidenceType.TEXT,
    payload: Buffer.from("Rollback scenario: system blocks UNKNOWN required predicate."),
    meta: { name: "rollback_test_log", source: "demo" },
  });

  const evidenceIndex = reg.exportIndex();

  console.log("\n=== Evidence Registered ===");
  console.log("Evidence Root Hash:", evidenceIndex.rootHash);
  console.log("Evidence IDs:");
  console.log(" -", ev1.evidenceId, (ev1.meta as Record<string, unknown>).name);
  console.log(" -", ev2.evidenceId, (ev2.meta as Record<string, unknown>).name);

  // --- Build juror prompt ---
  const prompt = buildJurorPrompt({
    role: "GENERAL_JUROR",
    caseId,
    predicates: requiredPredicates,
    evidenceIndex: evidenceIndex as unknown as Record<string, unknown>,
  });

  console.log("\n=== Juror Prompt (paste this into your LLM juror) ===\n");
  console.log(prompt);

  // --- Collect juror outputs (manual paste) ---
  const jurors: JurorInput[] = [];

  console.log("\n=== Paste juror outputs ===");
  console.log("You can provide 1+ jurors. Leave juror_id empty to stop.\n");

  while (true) {
    const jurorId = (await ask("Juror ID (e.g. gpt, gemini, human1) [enter to finish]: ")).trim();
    if (!jurorId) {
      break;
    }
    const rawText = await readMultiline(`\nPaste output for juror '${jurorId}':`);
    jurors.push({ jurorId, rawText });
    console.log();
  }

  if (jurors.length === 0) {
    console.log("\nNo juror outputs provided. Exiting.");
    rl.close();
    return;
  }

  // --- Run session ---
  const result = runSession({
    caseId,
    requiredPredicates,
    jurorInputs: jurors,
    evidenceRegistry: reg,
    signerKid,
    signerKey,
    auditLogPath: auditPath,
    forceEscalate: false,
  });

  console.log("\n=== Session Result ===");
  console.log("Verdict:", result.verdict);
  console.log("Blocking predicates:", result.blockingPredicates);
  console.log("Predicate results:");
  for (const [k, v] of Object.entries(result.predicateResults)) {
    console.log(` - ${k}: ${v}`);
  }

  console.log("\n=== Signed Receipt ===");
  console.log("Unsigned output (what was signed):");
  
  // Reconstruct the unsigned payload from known fields (snake_case to match signed format)
  const unsignedPayload = {
    schema_version: "jqs.output.v0.7",
    case_id: caseId,
    verdict: result.verdict,
    blocking_predicates: result.blockingPredicates,
    predicate_results: result.predicateResults,
    evidence_registry_root: reg.rootHash(),
    diagnostics: {
      jurors_total: result.diagnostics.jurorsTotal,
      jurors_parsed: result.diagnostics.jurorsParsed,
      jurors_empty: result.diagnostics.jurorsEmpty,
      citation_enforcement: result.diagnostics.citationEnforcement,
    },
  };
  console.log(JSON.stringify(unsignedPayload, null, 2));
  console.log("\nreceipt_sig:", result.receipt);

  // --- Verify receipt (self-check) ---
  const ok = verifyReceipt(
    unsignedPayload,
    result.receipt as { kid: string; sig_hex: string },
    new Map([[signerKid, signerKey]])
  );
  console.log("\nReceipt verifies:", ok);

  // --- Write files for independent verification ---
  const outputDir = "output";
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  const unsignedPath = path.join(outputDir, "unsigned.json");
  const receiptPath = path.join(outputDir, "receipt_sig.json");
  const keysPath = path.join(outputDir, "keys.json");

  fs.writeFileSync(unsignedPath, JSON.stringify(unsignedPayload, null, 2));
  fs.writeFileSync(receiptPath, JSON.stringify(result.receipt, null, 2));
  fs.writeFileSync(keysPath, JSON.stringify({
    [signerKid]: { encoding: "utf8", value: "dev-secret-change-me" }
  }, null, 2));

  console.log("\n=== Output Files (for independent verification) ===");
  console.log("Written:");
  console.log(" -", unsignedPath);
  console.log(" -", receiptPath);
  console.log(" -", keysPath);
  console.log("\nTo verify independently:");
  console.log(`  npx tsx tools/verify-receipt.ts ${unsignedPath} ${receiptPath} ${keysPath}`);

  console.log("\nAudit log written to:", auditPath);
  console.log("Done.");

  rl.close();
}

main().catch((err) => {
  console.error("Error:", err);
  process.exit(1);
});
